### PR TITLE
Add word to number function. 

### PR DESCRIPTION
**Pull request by kianenigma**
fixes https://github.com/rustland-fa/persian-tools-rs/issues/18

I am still not 100% sure if the logic is correct and there are no edge cases. Like it or not, I didn't really look at the `ts` example, and started off totally fresh to see what I come up with. I won't be shocked if there are some stupid mistakes in there :D 

A few TODOs left intentionally in the code to raise attention in review. WDYT? 

Lastly, typing Farsi in vscode was not fun, not fun. Yes, I am talking to you https://github.com/microsoft/vscode/issues/86667

### CODE REVIEW

File: src/digit/mod.rs [hunk 1]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+            .into_iter()
+            .filter(|t| (*t).ne("و"))
+            .map(parse)
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut final_value: N = num_traits::Zero::zero();
+        let mut intermediary_value: N = num_traits::Zero::zero();
+
+        let check_add = |this: N, that: N| this.checked_add(&that).ok_or(CANT_CONVERT);
+
```
Comment (kianenigma) [inline src/digit/mod.rs:174]:
```suggestion
```

File: src/digit/mod.rs [hunk 2]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+    "ده" => 10,
+    "یازده" => 11,
+    "دوازده" => 12,
+    "سیزده" => 13,
+    "چهارده" => 14,
+    "پانزده" => 15,
+    "شانزده" => 16,
+    "هفده" => 17,
+    "هجده" => 18,
+    "نانزده" => 19,
```
Comment (omid) [inline src/digit/mod.rs:93]: احتمالن «نوزده» صحیح‌تر باشه :)

https://fa.wikipedia.org/wiki/%DB%B1%DB%B9_(%D8%B9%D8%AF%D8%AF)

File: src/digit/mod.rs [hunk 3]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+        // 1 multiplier
+        assert_eq!("بیست و یک".words_to_number::<u32>().unwrap(), 21);
+        assert_eq!("نود و نه".words_to_number::<u32>().unwrap(), 99);
+        assert_eq!("دویست و هشت".words_to_number::<u32>().unwrap(), 208);
+
+        assert_eq!("هزار و یازده".words_to_number::<u32>().unwrap(), 1011);
+        assert_eq!("یک هزار و یازده".words_to_number::<u32>().unwrap(), 1011);
+
+        assert_eq!(
+            "میلیون و یازده".words_to_number::<u32>().unwrap(),
```
Comment (omid) [inline src/digit/mod.rs:252]: Persian is buggy here, I never thought about this... I think, "Million o yaazdah" is wrong, but "Hezar o yaazdah" is correct 🤔 

I think we can leave this bug to the language, not our code :D

File: src/digit/mod.rs [hunk 4]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+    "میلیون" => 1_000_000,
+    "میلیارد" => 1_000_000_000,
+};
+
+/// Fixed numbers that we interpret at face-value, as-is, because they cannot be broken down into
+/// smaller parts.
+///
+/// Includes [1-20], [30, 40, ..., 100], and [100, 200, ..., 900]
+// TODO: probably move to another file, too much bloat here.
+// TOOD: Is it 'nohsad' or 'noh sad'? 'haftsad or 'haft sad'?
```
Comment (omid) [inline src/digit/mod.rs:72]: Or 'haft[ZWNJ]sad' :))

File: src/digit/mod.rs [hunk 5]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut final_value: N = num_traits::Zero::zero();
+        let mut intermediary_value: N = num_traits::Zero::zero();
+
+        let check_add = |this: N, that: N| this.checked_add(&that).ok_or(CANT_CONVERT);
+
+        let checked_mul = |this: N, that: N| this.checked_mul(&that).ok_or(CANT_CONVERT);
+
+        for t in parsed {
```
Comment (omid) [inline src/digit/mod.rs:177]: I think we need an extra check here in this loop.
For example, linguistically, we cannot have two "Hezar"s in one number, but here we can.
So "صد و یک هزار و بیست هزار و شش هزار" will produce 127,000.

File: src/digit/mod.rs [hunk 6]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+        let parsed = tokens
+            .into_iter()
+            .filter(|t| (*t).ne("و"))
+            .map(parse)
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut final_value: N = num_traits::Zero::zero();
+        let mut intermediary_value: N = num_traits::Zero::zero();
+
+        let check_add = |this: N, that: N| this.checked_add(&that).ok_or(CANT_CONVERT);
```
Comment (omid) [inline src/digit/mod.rs:173]: Do we really need to check for an overflow for addition and multiplication?
Currently, the maximum number can be 999_999_999_999.
Seems like it's not possible to have an overflow.

Comment (omid) [reply src/digit/mod.rs:173]: Ah, OK, N is coming from the function call.
Maybe we can just use i128 instead?

**Comment by omid** [comment]
We can also add the support of "Manfi/Mosbat" to produce negative numbers.

File: src/digit/mod.rs [hunk 7]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
 impl Digit for String {}
 impl Digit for str {}
 
+/// The multipliers of the persian number system, up to a billion.
+pub static MULTIPLIERS: phf::Map<&'static str, u32> = phf::phf_map! {
```
Comment (omid) [inline src/digit/mod.rs:61]: I think nobody outside of this library needs these hashmaps, maybe we can make them private.

File: src/digit/mod.rs [hunk 8]
```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
...
+        // any gibberish fails
+        assert!("صد و و کیان هزار و بیست و شش"
+            .words_to_number::<u32>()
+            .is_err());
+    }
+
+    #[test]
+    fn boundaries() {
+        // boundaries
+        // assert_eq!("صفر".words_to_number::<u32>().unwrap(), 0);
```
Comment (omid) [inline src/digit/mod.rs:297]: A poor forgotten commented test case :D

**Comment by omid** [comment]
The TS library has the support of typo correction with the levenshtein algorithm.
We can also have similar functionality here.

https://crates.io/crates/levenshtein

**Comment by robatipoor** [comment]
As Omid said, this code needs some changes that I will do myself  thank you all

**Comment by kianenigma** [comment]
Okay, sorry for my delays. I usually only spend Sundays on such projects, so I was planning to address some of the comments then, but yeah feel free to pick it up as well. I'll start working on sth else next Sunday 😄

**Pull request merged**